### DISCUSSION TITLE
Possible to render different images to each eye?

### DISCUSSION THREAD

**Discussion by pj6** [StereoKit/StereoKit#798] [Q&A] [answered]
Just want to display two equirectangular 360 images to make a basic 3D image viewer as a test, couldn't find direct support for this in the source or samples.

**Comment by diniramdhni** [comment]


**Reply by maluoi** [reply]
...is this a ChatGPT copy-paste?

**Comment by maluoi** [comment, marked as answer]
So, in a shader you do kinda have access to the eye! But currently this only works on platforms where I have single-pass rendering working, which is just Windows Desktop and HoloLens. If that's enough for you, this trick will work, but if not, you might need to get a bit creative. I'll look into adding something that can enable this even without single-pass rendering, but it's not there today.

The key thing here is that shaders have access to a view id in the vertex shader. Here's how every SK vertex shader starts off:

```HLSL
psIn vs(vsIn input, uint id : SV_InstanceID) {
	psIn o;
	o.view_id = id % sk_view_count;
	id        = id / sk_view_count;
```
`o.view_id` here is the eye index (only in single-pass, unfortunately the multipass fallback uses an sk_view_count of 1, and calls this all twice). You can then use this id to switch between two different textures.

**Reply by pj6** [reply]
It's weird that this isn't simple. It's just like when I first looked at DirectX and wanted to draw a flat image, and learned I had to make 3D geometry and make a "texture" with it. I didn't want the 3D engine but I was forced to use it. You'd think 2D would be the fundamental operation since in the end all display hardware is flat, but looks like this hasn't changed.

Actually got StereoKitC to compile, I might be able to figure it out. Will the whole render have to be single pass or can I mix modes? That is assuming I can figure out how to put it into single pass.

**Reply by maluoi** [reply]
Single-pass vs. multi-pass is currently an engine implementation detail. I've figured out how to do it with DirectX, but not OpenGL yet. So you get single-pass on DirectX platforms like Windows and HoloLens, and multi-pass elsewhere.

I'd love to make this simpler though, I'll try and write a sample for this soon :)
